### PR TITLE
throtting: implemented a Throttler based on token bucket and configur…

### LINKED ISSUE #1502: Implement upload and download throttling using wrapper provider
Currently upload and download throttling is done at the provider level, which means it's not uniformly supported. For example filesystem, webdav and sftp do not support it, while all 4 cloud providers do support it.

We can make this more uniform by implementing a wrapper provider (like `readonly`, `logging`, `retrying`, etc.), which would do rate limiting in terms of bytes sent and received before/after sending the request to the underlying provider.

### PR DESCRIPTION
**Pull request by jkowalski**
This unifies throttling for all providers.

Fixes #1502 
Related #1503

### CODE REVIEW

File: repo/blob/throttling/throttler_test.go [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+		}
+
+		wg.Wait()
+
+		numSeconds := timer.Elapsed().Seconds()
+		actualRate := float64(*total) / numSeconds
+
+		// make sure the rate is less than target with some tiny margin of error
+		require.Less(t, actualRate, wantRate*1.05)
+		require.Greater(t, actualRate, wantRate*0.9)
```
Comment (julio-lopez) [inline repo/blob/throttling/throttler_test.go:133]: will this be flaky?

File: repo/blob/throttling/token_bucket.go [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+type tokenBucket struct {
+	name  string
+	now   func() time.Time
+	sleep func(ctx context.Context, d time.Duration)
+
+	mu                   sync.Mutex
+	lastTime             time.Time
+	numTokens            float64
+	maxTokens            float64
+	addTokensPerTimeUnit float64
```
Comment (julio-lopez) [inline repo/blob/throttling/token_bucket.go:24]: is there any case where `addTokensPerTimeUnit != maxTokens` ?

**Review by julio-lopez — COMMENTED** [review]
+1: LG 👍

File: repo/blob/throttling/throttler_test.go (hunk 1 above)
Comment (jkowalski) [reply repo/blob/throttling/throttler_test.go:133]: i ran it enough times - 5% buffer appears sufficient but we can raise it if it flakes out.

File: repo/blob/throttling/token_bucket.go (hunk 2 above)
Comment (jkowalski) [reply repo/blob/throttling/token_bucket.go:24]: not really, removed for now

**Pull request merged**